scylla-cdc/src/reader.rs

```diff
@@ -0,0 +1,266 @@
+                )
+                .await?
+                .into_typed::<(i32, i32, String, String)>();
+
+            while let Some(row) = rows_stream.next().await {
+                let (pk, t, v, s) = row?;
+                println!("pk: {} t: {} v: {} s: {}", pk, t, v, s);
+            }
+
+            lower_timestamp = upper_timestamp;
+            sleep(self.sleep_interval).await;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use scylla::query::Query;
+    use scylla::statement::Consistency;
+    use scylla::SessionBuilder;
+    use std::error::Error;
+
+    const TEST_KEYSPACE: &str = "Test";
+    const TEST_TABLE: &str = "t";
+    const SLEEP_INTERVAL: u64 = 1;
+    const WINDOW_SIZE: i64 = 2;
+    const SAFETY_INTERVAL: i64 = 1;
+
+    impl Reader {
+        fn test_new(
+            stream_ids: Vec<StreamID>,
+            start_timestamp: chrono::Duration,
+            window_size: chrono::Duration,
+            safety_interval: chrono::Duration,
+            sleep_interval: time::Duration,
+        ) -> Reader {
+            Reader {
+                stream_id_vec: stream_ids,
+                lower_timestamp: start_timestamp,
+                window_size,
+                safety_interval,
+                upper_timestamp: None,
+                sleep_interval,
+            }
+        }
+    }
+
+    async fn setup(session: &Session) -> Result<Reader, Box<dyn Error>> {
+        create_test_db(session).await;
+        populate_test_db(session).await;
+        let stream_id = get_cdc_stream_id(session).await?;
+
+        let start_timestamp: chrono::Duration =
+            chrono::Duration::from_std(SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?)?;
+        let sleep_interval: time::Duration = time::Duration::new(SLEEP_INTERVAL, 0);
+        let window_size: chrono::Duration = chrono::Duration::seconds(WINDOW_SIZE);
+        let safety_interval: chrono::Duration = chrono::Duration::seconds(SAFETY_INTERVAL);
+
+        let reader = Reader::test_new(
+            Vec::from([stream_id]),
+            start_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+        );
+
+        Ok(reader)
+    }
+
+    fn get_create_table_query() -> String {
+        format!("CREATE TABLE {}.{} (pk int, t int, v text, s text, PRIMARY KEY (pk, t)) WITH cdc = {{'enabled':true}};",
+                TEST_KEYSPACE,
+                TEST_TABLE
+        )
+    }
+
+    async fn create_test_db(session: &Session) {
+        let mut create_keyspace_query = Query::new(format!(
+            "CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = {{'class': 'SimpleStrategy', 'replication_factor': 1}};",
+            TEST_KEYSPACE
+        ));
+        create_keyspace_query.set_consistency(Consistency::All);
+        session.query(create_keyspace_query, &[]).await.unwrap();
+
+        // Create test table
+        let create_table_query = get_create_table_query();
```

Comment:
This will fail if the table already exists. You should add `IF NOT EXISTS` to your query or make sure you tear everything down after the tests are run.

Reply:
To clarify: `delete_test_data_in_db` is not called when your tests fail (for example when some database or networking error happens - simply provide invalid database URI and you'll experience it) leaving the test database in a messy state.

Reply:
You can look at the tests in `stream_generations.rs` for inspiration on how to initialize the database for the tests: https://github.com/piodul/scylla-cdc-rust/blob/1974fbf3b1ab0b17048835674424560e1af49fff/scylla-cdc/src/stream_generations.rs#L301-L331